src/github-access.tsx

```diff
@@ -0,0 +1,99 @@
+import * as http from 'http';
+import * as url from 'url';
+import * as open from 'open';
+import { saveEnvVariable } from './variables';
+import { fetch } from 'cross-fetch';
+
+const GITHUB_ACCESS_TOKEN = 'Github__OAuthAccessToken';
+
+type OAuthReturn = {
+  access_token: string;
+  scope: string;
+  token_type: string;
+};
+
+export class GithubAccess {
+  // "avensia-oss-garn" oauth app
+  private static OAuthClientId = '825e1cb8ef39aa07f127';
+  // This is safe to store in version control
+  private static OAuthClientSecret = 'a7e9440c087ad4410dc4f77485622cc7a17feda3';
+  // This port is important. It's defined as a valid callback url in the app settings in github
+  private static capturePort = 9437;
+
+  public static get accessToken() {
+    return process.env[GITHUB_ACCESS_TOKEN] || null;
+  }
+
+  public static set accessToken(token: string | null) {
+    process.env[GITHUB_ACCESS_TOKEN] = token ?? '';
+    saveEnvVariable(GITHUB_ACCESS_TOKEN, token ?? '');
+  }
+
+  public static hasCredentials(): boolean {
+    return !!GithubAccess.accessToken;
+  }
+
+  public static async validateCredentials() {
+    if (!GithubAccess.hasCredentials()) {
+      try {
+        await GithubAccess.captureCredentials();
+      } catch (e) {
+        throw new Error(`Failed to capture github credentials: ${e}`);
+      }
+    }
+  }
+
+  private static async captureCredentials() {
+    const promise = new Promise((resolve, reject) => {
+      const nonce = String(Math.random()).split('.')[1];
+      const requestAuthUrl = `https://github.com/login/oauth/authorize?client_id=${GithubAccess.OAuthClientId}&state=${nonce}&scope=repo`;
+
+      // Open browser tab which prompts user to authorize the garn app.
+      open(requestAuthUrl);
+
+      const server = http.createServer(async (req, res) => {
+        const queryObject = url.parse(req.url!, true).query;
+        const code = queryObject.code;
+        const state = queryObject.state;
+
+        if (code) {
+          if (state === nonce) {
+            res.statusCode = 200;
+            res.setHeader('content-type', 'text/html');
+            res.end('Validation OK! You can now close this tab. <script>setTimeout(() => close(), 1000)</script>');
+
+            server.close();
+            // Fetch credentials using temporary code.
+            const authUrl = `https://github.com/login/oauth/access_token?client_id=${GithubAccess.OAuthClientId}&client_secret=${GithubAccess.OAuthClientSecret}&code=${code}`;
```

Comment:
i don't agree it's safe to include client secret which is distributed to everyone. we should use this oauth flow https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce

Reply:
I agree that it looks weird to include the client secret. I followed github's [guide for authorizing oauth apps](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps), with inspiration from the auth flow in [github's cli](https://github.com/cli/cli/blob/6a8deb1f5a9f2aa0ace2eb154523f3b9f23a05ae/internal/authflow/flow.go#L25).

Reply:
Changed the auth flow to use device-auth instead of client_secret!